src/operation_manager.rs

```diff
@@ -0,0 +1,530 @@
+        let mut should_interpret = false;
+
+        // Find out which operations should be invalidated.
+
+        for (var_ident_id, selected_op) in self.selected_ops.iter_mut().enumerate() {
+            // First dirty operation marks all subsequent ones as invalidated.
+            if !invalidate_statements && selected_op.param_values_dirty() {
+                invalidate_statements = true;
+
+                log::info!("Operation \"{}\" is marked as dirty.", selected_op.op.name);
+
+                selected_op.consume_all_values();
+            }
+
+            if invalidate_statements {
+                let var_decl = build_operation_var_ident(&selected_op.op, var_ident_id as u64);
+
+                self.interpreter_server
+                    .submit_request(InterpreterRequest::SetProgStmtAt(var_ident_id, var_decl));
+
+                // If this operation wasn't even run, there is nothing to
+                // invalidate.
+                if selected_op.status == OpStatus::Finished {
+                    results_to_invalidate += 1;
+                }
+
+                should_interpret = true;
+                selected_op.status = OpStatus::Running;
+            }
+        }
+
+        log::info!(
+            "Geometries from {} operations will be invalidated.",
+            results_to_invalidate
+        );
+
+        let mut geometries_to_invalidate: usize = 0;
+
+        // Find out which geometries should be invalidated.
+
+        for _ in 0..results_to_invalidate {
+            let result_geometries = self
+                .geometry_stack
+                .pop()
+                .expect("Failed to pop value off geometry stack");
+
+            for _ in 0..result_geometries {
+                let geometry_metadata = self
+                    .geometry_metadata
+                    .pop()
+                    .expect("Failed to pop value off geometry metadata");
+
+                // Since frontend shows only unused geometries, only those
+                // should be counted.
+                if !geometry_metadata.used {
+                    geometries_to_invalidate += 1;
+                }
+            }
+        }
+
+        if should_interpret {
+            self.interpreter_server
+                .submit_request(InterpreterRequest::Interpret);
+        }
+
+        log::info!(
+            "{} geometries can be removed from the scene.",
+            geometries_to_invalidate
+        );
+
+        geometries_to_invalidate
+    }
+
+    /// Polls for respone from interpreter server and runs the handler function
+    /// for each unused geometry in case a result is returned.
+    ///
+    /// Since unused geometries are the only ones to be shown in scene, the
+    /// handler function runs only for them since the geometry ID needs to be
+    /// assigned.
+    pub fn poll_interpreter_response<F>(&mut self, mut geometry_handler: F)
+    where
+        F: FnMut(&mut GeometryMetadata) -> GpuGeometryId,
```

Comment:
As noted in the main review comment and also elsewhere, it would seem better if instead of using this to inject renderer ids, we would generate our own value ids here and publish them from this function. The caller code could then store the mapping between values and renderer geometries.